### PR TITLE
Adding Crosscompiling section, Example Dockerfile

### PR DESCRIPTION
**Pull request by xristos-sk**
In this PR : 

- Creating a new Page under **4. Development** about crosscompiling `cargo-flash` by using `cross` or `LLVM`
- Adding an Example Dockerfile for `cross` usage

I had an issue with building `cargo-flash` on the **Rasp pi 400**, as it went OOM on the `rustc` linking step.

The Probe-rs Matrix channel was very helpfull and pointed me into the direction of `cross` and `LLVM` targets as options to succesfully overcome this limitation, including an example `Dockerfile`. Now i want to share the knowledge.

While writing the documentation for `LLVM` though, it became apparent that it would require a lof of information about how to use and debug it's toolchain, which made the page go out of context ( in my opinion ). So a link to how someone can use `LLVM` to Crosscompile replaced that section.

In the way i see it, this PR satisfies the initial goal, as any other user that experiences my issue on compiling in limited environments will now have somewhere to start, and a solution using `cross`. 

in short, If you already know how to use LLVM for a different target, you will do it. if not, learning how might be out of the scope of this page. Open for discussion on this point.

### CODE REVIEW

File: content/guide/4_development/crosscompiling.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+git clone https://github.com/probe-rs/cargo-flash
+
+# Installing cross
+cargo install cross
+
+# CD into the cargo-flash repository
+# And Create a `crossimage` folder
+cd cargo-flash && mkdir crossimage
+
+# Create a Dockerfile resempling the target system,
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:35]:
```suggestion
# Create a Dockerfile resembling the target system,
```

File: content/guide/4_development/crosscompiling.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+> “Zero setup” cross compilation and “cross testing” of Rust crates
+
+It's main advantage is to **not be suseptible** on the usual cross compilation issues that arise **from using Host machine's shared object libraries**, by using a Docker image to build the static binary.
+
+The downside of this approach is that **you will have to rely on the Cross Project** and **supply the build environment Dockerfile**, making sure that it resembles as much as possible the system you are building for with **a cross C toolchain installed**.
+
+Building `cargo-flash` for the Raspberry Pi 400, on an `x86_64` Linux System will happen as follows :
+
+```sh
+# Cloning cargo-flash on the system we will build the binary
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:25]:
```suggestion
# Cloning cargo-flash on the system we will build the binary on.
```

File: content/guide/4_development/crosscompiling.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+git clone https://github.com/probe-rs/cargo-flash
+
+# Installing cross
+cargo install cross
+
+# CD into the cargo-flash repository
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:31]:
```suggestion
# `cd` into the cargo-flash repository
```

File: content/guide/4_development/crosscompiling.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+# And Create a `crossimage` folder
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:32]:
```suggestion
# And create a `crossimage` folder
```

File: content/guide/4_development/crosscompiling.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+cd cargo-flash && mkdir crossimage
+
+# Create a Dockerfile resempling the target system,
+# and put it inside the crossimage Folder
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:36]:
```suggestion
# and put it inside the `crossimage` folder.
```

File: content/guide/4_development/crosscompiling.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+We can take a `armv7-unknown-linux-gnueabihf` base system from the `rustembedded/cross` Docker hub container registry.
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:39]:
```suggestion
We can take an `armv7-unknown-linux-gnueabihf` base system from the `rust-embedded/cross` Docker Hub container registry.
```

File: content/guide/4_development/crosscompiling.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+[Cross](https://github.com/rust-embedded/cross) self defines as:
+
+> “Zero setup” cross compilation and “cross testing” of Rust crates
+
+It's main advantage is to **not be suseptible** on the usual cross compilation issues that arise **from using Host machine's shared object libraries**, by using a Docker image to build the static binary.
+
+The downside of this approach is that **you will have to rely on the Cross Project** and **supply the build environment Dockerfile**, making sure that it resembles as much as possible the system you are building for with **a cross C toolchain installed**.
+
+Building `cargo-flash` for the Raspberry Pi 400, on an `x86_64` Linux System will happen as follows :
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:22]:
```suggestion
Building `cargo-flash` for the Raspberry Pi 400, on an `x86_64` Linux System happens as follows:
```

File: content/guide/4_development/crosscompiling.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+# And Create a `crossimage` folder
+cd cargo-flash && mkdir crossimage
+
+# Create a Dockerfile resempling the target system,
+# and put it inside the crossimage Folder
+```
+
+We can take a `armv7-unknown-linux-gnueabihf` base system from the `rustembedded/cross` Docker hub container registry.
+
+Then, we need to follow the instructions to prepare it for building steps the image from the `probe-rs`, `cargo-flash` [prerequisites](https://github.com/probe-rs/cargo-flash#prerequisites).
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:41]:
```suggestion
After that, follow the instructions from the `probe-rs`, `cargo-flash` [prerequisites](https://github.com/probe-rs/cargo-flash#prerequisites) to prepare the image for the building steps.
```

File: content/guide/4_development/crosscompiling.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+You can download the Dockerfile for this example from [here](/content/cross-dockerfile.txt)
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:43]:
```suggestion
You can download the Dockerfile for this example from [here](/content/cross-dockerfile.txt).
```

File: content/guide/4_development/crosscompiling.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+```sh
+# Create and edit a Cross.toml file on the root of the cloned repo
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:46]:
```suggestion
# Create and edit a Cross.toml file in the root of the cloned repo.
```

File: content/guide/4_development/crosscompiling.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+vim Cross.toml
+```
+
+Add the following inside the **Cross.toml**. Which will define to `cross` which container should be used for the target Architecture :
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:50]:
```suggestion
Add the following to the `Cross.toml`. Which will define to `cross` which container should be used for the target architecture:
```

File: content/guide/4_development/crosscompiling.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+<pre>[target.armv7-unknown-linux-gnueabihf]
+image = "crossimage"</pre>
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:53]:
```suggestion
    ```toml
    [target.armv7-unknown-linux-gnueabihf]
    image = "crossimage"
    ```
```

File: content/guide/4_development/crosscompiling.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+```sh
+# Build and tag the container image, specifying the name as defined on Cross.toml
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:56]:
```suggestion
# Build and tag the container image, specifying the name as defined in the `Cross.toml`.
```

File: content/guide/4_development/crosscompiling.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+docker build -t crossimage crossimage/
+
+# Run cross to compile, cross arguments are the same as the `cargo` ones
+cross build --release --target=armv7-unknown-linux-gnueabihf
+
+# Done
+```
+
+You will now have a statically built `cargo-flash` binary inside the `target/<architecture>/release` folder on the root directory of the cloned repo, which you can now move to the **Raspberry Pi 400** `Cargo bin Path` folder ( usually `~/.cargo/bin` ).
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:65]:
```suggestion
You now have a statically built `cargo-flash` binary inside the `target/<architecture>/release` folder on the root directory of the cloned repo, which you can now move to the **Raspberry Pi 400** `cargo bin path` folder (usually `~/.cargo/bin`).
```

File: content/guide/4_development/crosscompiling.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+## Using LLVM
+
+As using `LLVM` for cosscompiling can go out of the scope of just generating a `cargo-flash` binary, Please follow the instructions on how to Crosscompile from the **LLVM** [guide](https://www.llvm.org/docs/HowToCrossCompileLLVM.html).
```
Comment (Yatekii) [inline content/guide/4_development/crosscompiling.md:69]:
```suggestion
As using `LLVM` for cosscompiling can go out of the scope of just generating a `cargo-flash` binary, please follow the instructions on how to crosscompile in the **LLVM** [guide](https://www.llvm.org/docs/HowToCrossCompileLLVM.html).
```

**Review by Yatekii — CHANGES_REQUESTED** [review]
Hey, thanks a lot for doing this!

Sorry for the long delay.

I finally proof read everything. Would you be so kind and fix the pointed out things?

**Comment by xristos-sk** [comment]
Sorry for the actions spam :) i eventually batched them together.

All good, thank you for making sure my broken English are not going to appear upstream.
